freqtrade/optimize/__init__.py

```diff
@@ -15,6 +15,63 @@
 logger = logging.getLogger(__name__)
 
 
+def sanitize_start_stop_date(tickerlist: List[Dict], timerange, pair):
+    """
+    This function is called when either timerange start or stop are type date
+    -
+    Prevent index out or range error in trim_tickerlist, which occurs when trying
+    to process a stop or start range outside that in the cached data
+    returns (start, stop, nothing_to_trim)
+    nothing_to_trim  > 0 is a flag to later not send the pair data to trim_tickerlist
+    -
+    Function Logic:
+     - if cache records begin after timerange start - reset start to first record
+     - if cache records end before timerange stop - reset stop to last record
+     - if start or stop are fully after or before cache records - return, nothing_to_trim.
+     - if only 1 of stop or start are set in timerange use the first or last
+       record appropriately from the cache as the unset value
+    """
+
+    # Do nothing if timerange does not contain a type date.
+    stype, start, stop = timerange
+    nothing_to_trim = 0
+    if stype[0] != 'date' or stype[1] != 'date':
+        sanitized_dates = [start, stop, nothing_to_trim]
+        return sanitized_dates
+
+    # If no arg for stop or start then set to the first or last record in the cache
+    stop = stop if stype[1] == 'date' else int(((tickerlist[(len(tickerlist) - 1)][0]) / 1000))
+    start = start if stype[0] == 'date' else int(((tickerlist[0][0]) / 1000))
+
+    # If requested range start is after cache records end - no data to be trimmed
+    # If requested start range is before start of cache, move it to first nearest record
+    if stype[0] == 'date':
+        if (tickerlist[0][0]) > (stop * 1000):
+            logger.warn('No data for %s timerange in cache, update cache ', pair)
+            nothing_to_trim = nothing_to_trim + 1
+        elif (tickerlist[0][0]) > (start * 1000):
+            start = (tickerlist[0][0] / 1000)
+            logger.warn('Requested start timerange for %s not in cache, update cache ', pair)
+
+    # If requested range stop is before cache records begin - no data to be trimmed
+    # If requested stop range is after end of cache, move it to last nearest record
+    if stype[1] == 'date':
+        if (tickerlist[(len(tickerlist) - 1)][0]) < (start * 1000):
+            logger.warn('No data for %s timerange in cache, update cache ', pair)
+            nothing_to_trim = nothing_to_trim + 2
+        elif (tickerlist[(len(tickerlist) - 1)][0]) < (stop * 1000):
+            stop = (tickerlist[(len(tickerlist) - 1)][0] / 1000)
+            logger.warn('Requested stop timerange for %s not in cache, update cache ', pair)
+
+    # Impossible range, nothing to trim.
+    if start > stop:
```

Comment:
This case should honestly be cought on initialisation - calling with start > stop should cause backtesting to stop immediately ... 
I would add testing if start is later than stop to the below function in `arguments.py`
https://github.com/gcarq/freqtrade/blob/develop/freqtrade/arguments.py#L214

Reply:
The challenge here is we may change start/stop date when adjusting to the first/last record available from the json cache.  I think its a good sanity check after the change function - i could make the comment better